### PR TITLE
Performance/accuracy: expm1

### LINKED ISSUE #45046: [Bounty $10000] Optimise/improve accuracy for expm1 (fp32/bf16)
Current performance (measured on WH):

- 99 cycles per iteration for fp32; >1.6 ulp
- 47 cycles per iteration for bf16; >1.2 ulp

Performance/accuracy targets:

- 66 cycles (1.5x) for fp32; <1 ulp
- 45 cycles (1.04x) for bf16; <0.5 ulp

### PR DESCRIPTION
**Pull request by jasondavies**
### PR Category

Performance/accuracy

### Summary

Closes https://github.com/tenstorrent/tt-metal/issues/45046

| type | old cycles | new cycles | speedup | old maxulperr | new maxulperr |
|---|---:|---:|---:|---:|---:|
| fp32 | 99 | 52 | 1.90x | 1.6350 | 0.9974 |
| bf16 | 47 | 34 | 1.38x | 1.2120 | 0.4997 |

### Notes for reviewers

The `half` template parameter is there specifically for full-range input support for sinh (will be added in a separate PR).

### CODE REVIEW

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h [hunk 1]
```diff
@@ -1,149 +1,207 @@
...
+        sfpi::vFloat w = 0.5f;
+        r = __builtin_rvtt_sfpmad(r.get(), f.get(), w.get(), sfpi::SFPMAD_MOD1_OFFSET_NONE);
 
-// is_fp32_dest_acc_en == true: use float32-accurate version
-template <>
-sfpi_inline sfpi::vFloat _sfpu_expm1_improved_<true>(sfpi::vFloat val) {
-    return _sfpu_expm1_f32_accurate_(val);
+        sfpi::vFloat infinity = std::numeric_limits<float>::infinity();
+
+        r = r * s + f;
```
Comment (Aswinmcw) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h:91]: The `v_block` + `v_and(j != 0.0f)` pattern for the i==0 fast path is elegant — avoids branching cost for the common case where no scaling is needed.

File: tests/ttnn/unit_tests/operations/eltwise/test_expm1.py [hunk 2]
```diff
@@ -43,9 +39,11 @@ def test_expm1_arange_masking(dtype, device):
...
-    # masking to working range
-    mask = (input_tensor >= low) & (input_tensor <= high)
-    input_tensor = input_tensor[mask]
+    # BF16 NaN values are packed as infinity during device transfer, so they are not
+    # valid inputs for checking expm1's NaN propagation. The float32 parametrisation
+    # covers NaN propagation using the same generated bit patterns.
+    if tt_dtype == ttnn.bfloat16:
+        input_tensor = input_tensor[~torch.isnan(input_tensor)]
 
     tt_in = ttnn.from_torch(
```
Comment (Aswinmcw) [inline tests/ttnn/unit_tests/operations/eltwise/test_expm1.py:48]: Good catch filtering bf16 NaNs — they get packed as infinity during device transfer so they are not valid inputs for NaN propagation testing. The fp32 parametrisation still covers NaN cases.

**Review by Aswinmcw — APPROVED** [review]
Excellent rewrite. The new implementation replaces the hybrid Taylor/exp approach with Norbert Juffa's range-reduction method, which is fundamentally more accurate: it computes expm1(f) directly on the reduced range f = a - i*ln(2) without ever forming exp(x)-1.

Key improvements verified:
- **bf16**: removes the branching Taylor/exp split; single unified path via range reduction + polynomial. ULP tightened from 2 to 1.
- **fp32**: uses double-float ln(2) splitting (hi/lo) for precise range reduction, 6th-order polynomial for expm1(f), and careful reconstruction via 2*(r*t + t - 0.5) to avoid catastrophic cancellation.
- **Special cases**: +inf propagated via j*infinity multiply; large negative returns -1 (via -0.5 * 2); NaN propagated through multiply.
- **`half` template param**: clean API for callers needing expm1(x)/2 without an extra multiply.
- **Unified code path**: APPROXIMATION_MODE template parameter no longer selects different algorithms; both use the same implementation.
- **Test**: removes the artificial [low, high] masking (was hiding edge cases), filters bf16 NaNs (packed as inf during transfer), tightens ULP from 2 to 1.

The BSD-2 license from Norbert Juffa is properly attributed.

LGTM.

**Review by amahmudTT — APPROVED** [review]

**Comment by jasondavies** [comment]
@amahmudTT @fvranicTT Needs CI re-run.  Thanks!

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h [hunk 3]
```diff
@@ -1,149 +1,207 @@
...
- * This function implements expm1(x) = exp(x) - 1 with high accuracy for float32.
- * Target accuracy: < 2 ULP for float32.
+ * i = rint(a / log(2)), f = a - i * log(2). Then
+ * expm1(a) = 2**i * (expm1(f) + 1) - 1.
  *
- * Uses hybrid approach optimized for maximum ULP error:
- * - Taylor series (order 8) for |x| < 0.5 to avoid catastrophic cancellation
- * - exp(x) - 1 for |x| >= 0.5 (calls _sfpu_exp_fp32_accurate_)
+ * Compute r = expm1(f). Then
+ * expm1(a) = 2 * (0.5 * 2**i * r + 0.5 * 2**i - 0.5).
```
Comment (fvranicTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h:50]: What if we did:

```suggestion
 * expm1(a) = 2**i * r + (2**i - 1)..
```

That is, in fp16 path we change the v_block to look like:

```cpp
// For !half we encode 2**i directly so the scale/bias step produces
// expm1(a) without a trailing *2. For half we keep 0.5 * 2**i so the
// result is 0.5 * expm1(a).
sfpi::vFloat w_scale;
if constexpr (half) {
    w_scale = w;
} else {
    w_scale = sfpi::vConst1;
}
constexpr float overflow_neg = half ? -0.5f : -1.0f;

v_block {
    if constexpr (!half) {
        v_and(j != 0.0f);
    }
    sfpi::vFloat jm2 = j + -2.0f;
    scale = sfpi::reinterpret<sfpi::vFloat>((i << 23) + sfpi::reinterpret<sfpi::vInt>(w_scale));

    sfpi::vFloat abs_jm2 = sfpi::abs(jm2);
    bias = scale - w_scale;
    sfpi::vInt tail = sfpi::float_to_int8(abs_jm2, sfpi::RoundMode::NearestEven);
    r = scale * r + bias;

    v_if(tail >= 127) {
        // Positive side becomes +inf; NaNs should propagate through the multiply.
        r = jm2 * infinity;

        v_if(jm2 < 0.0f) { r = overflow_neg; }
        v_endif;
    }
    v_endif;
}
v_endblock;
```

Comment (jasondavies) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h:50]: Using 2**i directly can overflow; that's the reason for doing the reconstruction half-scaled.

Comment (fvranicTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h:50]: Also, now that I look again at the above code, is v_and part needed? 
When `j == 0`, `i = reinterpret_int(rounding_bias) = 0x4B400000`, and `(i << 23)` wraps to 0 in 32-bit arithmetic, giving `scale = w_scale` and `bias = 0`.
The FMA then produces `w_scale * r_poly`, which equals `0.5 * expm1(a)` (half=true) or `expm1(a)` (half=false), as expected when no exponent shift is needed.

If this holds, you don't need v_block at all, code becomes much simpler:

```cpp
sfpi::vFloat w_scale;
if constexpr (half) {
    w_scale = w;
} else {
    w_scale = sfpi::vConst1;
}
constexpr float overflow_neg = half ? -0.5f : -1.0f;

sfpi::vFloat jm2 = j + -2.0f;
scale = sfpi::reinterpret<sfpi::vFloat>((i << 23) + sfpi::reinterpret<sfpi::vInt>(w_scale));

sfpi::vFloat abs_jm2 = sfpi::abs(jm2);
bias = scale - w_scale;
sfpi::vInt tail = sfpi::float_to_int8(abs_jm2, sfpi::RoundMode::NearestEven);
r = scale * r + bias;

v_if(tail >= 127) {
    // Positive side becomes +inf; NaNs should propagate through the multiply.
    r = jm2 * infinity;

    v_if(jm2 < 0.0f) { r = overflow_neg; }
    v_endif;
}
v_endif;
```

Comment (fvranicTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h:50]: Ah, you are right. Thanks for pointing it out! Maybe capture that in a comment?

Comment (jasondavies) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h:50]: Removing the `j != 0.0f` check would unconditionally compute:

```
r = 0.5 * r_poly
r *= 2
```

Which underflows to zero for tiny values with raw exponent 1.

Comment (fvranicTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h:50]: In that case, my second argument about the need for v_and and v_block can be summed into:

```cpp
sfpi::vFloat jm2 = j + -2.0f;
scale = sfpi::reinterpret<sfpi::vFloat>((i << 23) + sfpi::reinterpret<sfpi::vInt>(w));

sfpi::vFloat abs_jm2 = sfpi::abs(jm2);
bias = scale - w;
sfpi::vInt tail = sfpi::float_to_int8(abs_jm2, sfpi::RoundMode::NearestEven);
r = scale * r + bias;

v_if(tail >= 127) {
    // Positive side becomes +inf; NaNs should propagate through the multiply.
    r = jm2 * infinity;

    v_if(jm2 < 0.0f) { r = -0.5f; }
    v_endif;
}
v_endif;
if constexpr (!half) {
    r *= 2.0f;
}
```

Comment (fvranicTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h:50]: Ok, then I'm out of ideas :)

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h [hunk 4]
```diff
@@ -1,149 +1,207 @@
...
+        r = r * f + sfpi::vConstFloatPrgm2;
+
+        sfpi::vFloat w = 0.5f;
+        r = __builtin_rvtt_sfpmad(r.get(), f.get(), w.get(), sfpi::SFPMAD_MOD1_OFFSET_NONE);
 
-// is_fp32_dest_acc_en == true: use float32-accurate version
-template <>
-sfpi_inline sfpi::vFloat _sfpu_expm1_improved_<true>(sfpi::vFloat val) {
-    return _sfpu_expm1_f32_accurate_(val);
+        sfpi::vFloat infinity = std::numeric_limits<float>::infinity();
```
Comment (fvranicTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h:89]: Would it matter if we hoist `infinity` out of both branches? Probably not?

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h (hunk 3 above)
Comment (jasondavies) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h:50]: Thanks for checking!  There's probably a separate question here of whether or not we need to support edge cases, particularly as Tensix is not IEEE-compliant which makes life much harder.  In this particular case, it's nice to cover the entire fp32 range of inputs, at the cost of a few cycles.  You could of course make an argument that the code is simpler if you drop the requirement for full range support.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h (hunk 4 above)
Comment (jasondavies) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h:89]: It would make it slower.  It's deliberately placed in both locations to hide the preceding SFPMAD latency.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h (hunk 3 above)
Comment (jasondavies) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_expm1.h:50]: Added some comments.

**Comment by jasondavies** [comment]
Merged with latest main to fix failing CI (ttsim-integration-tests `tests/ttnn/unit_tests/operations/conv/test_conv1d.py::test_conv1d_replicate_pad`)

**Comment by jasondavies** [comment]
@fvranicTT  Looks like a flaky runner?

```
  trigger-post-commit / ttsim-integration-tests / TTNN core ttnn unit test group - wormhole_b0

  The actual failure was an xdist worker crash:

  [gw1] node down: Not properly terminated
  FAILED tests/ttnn/unit_tests/base_functionality/test_reshape.py::test_reshape_oob
  worker 'gw1' crashed while running ...
```

**Comment by jasondavies** [comment]
@fvranicTT All green
